sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -63,15 +63,28 @@ case class Concat(children: Seq[Expression]) extends Expression with ImplicitCas
 
   override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
     val evals = children.map(_.genCode(ctx))
-    val inputs = evals.map { eval =>
-      s"${eval.isNull} ? null : ${eval.value}"
-    }.mkString(", ")
-    ev.copy(evals.map(_.code).mkString("\n") + s"""
-      boolean ${ev.isNull} = false;
-      UTF8String ${ev.value} = UTF8String.concat($inputs);
-      if (${ev.value} == null) {
-        ${ev.isNull} = true;
+    val argNums = evals.length
+    val args = ctx.freshName("argLen")
+    ctx.addMutableState("UTF8String[]", args, "")
+
+    val inputs = evals.zipWithIndex.map { case (eval, index) =>
+      if (eval.isNull != "true") {
+        s"""
+          ${eval.code}
+          if (!${eval.isNull}) {
+            $args[$index] = ${eval.value};
+          }
```

Comment:
I think we are better to assign null too if `eval.isNull == null`, because `args` is global variable and we need to override all values for previous row.

Reply:
Thanks, while `args` is global, `UTF8String[]` is allocated before executing these assignments. Thus, we can ensure all of elements in `args` are `null`.
What do you think?

Reply:
That is guaranteed before the first row processed. After we process rows, `args` are updated for each row. E.g., `args[0]` can be updated and assigned with a string for row 0. In next row, if `eval.isNull` is evaluated to true, we don't override back to null, so `arg[0]` is still the string value for row 0.

Reply:
When the next row is processed, `$argsnew UTF8String[$argNums]` is executed again in `apply()` method. In other words, current implementation does not reuse `UTF8String[]` between different rows.

Reply:
Oh, I see. I didn't notice that `args` is not globally initialized.

Reply:
Thank you for taking your time for my code.  
It is not easy to imagine the generated code. I may have to put the generated code.